File: src/server/src/velvet_dawn/mechanics/conditionals/conditional.py
import enum
from typing import Optional, Union

import velvet_dawn.validations
from velvet_dawn import errors
from velvet_dawn.dao.models import UnitInstance, TileInstance
from velvet_dawn.dao.models.world_instance import WorldInstance
from velvet_dawn.mechanics import selectors
from velvet_dawn.mechanics.function_value import FunctionValue
from velvet_dawn.mechanics.selectors import Selector


# Possible comparisons
class Comparison(enum.Enum):
    EQUALS = 0
    NOT_EQUALS = 1
    LESS_THAN = 2
    LESS_THAN_EQUAL = 3
    GREATER_THAN = 4
    GREATER_THAN_EQUAL = 5
    HAS_TAG = 6
    NOT_HAS_TAG = 7


# Comparison keys
OPERATORS = {
    "equals": Comparison.EQUALS, "eq": Comparison.EQUALS,
    "not-equals": Comparison.NOT_EQUALS, "ne": Comparison.NOT_EQUALS,
    "less-than": Comparison.LESS_THAN, "lt": Comparison.LESS_THAN,
    "less-than-equals": Comparison.LESS_THAN_EQUAL, "lte": Comparison.LESS_THAN_EQUAL,
    "greater-than": Comparison.GREATER_THAN, "gt": Comparison.GREATER_THAN,
    "greater-than-equals": Comparison.GREATER_THAN_EQUAL, "gte": Comparison.GREATER_THAN_EQUAL,
    "tagged": Comparison.HAS_TAG, "not-tagged": Comparison.NOT_HAS_TAG
}


# Numeric only operators
NUMBERS_ONLY_OPERATORS = {
    Comparison.LESS_THAN, Comparison.LESS_THAN_EQUAL, Comparison.GREATER_THAN_EQUAL, Comparison.GREATER_THAN
}


class Conditional:

    def __init__(self, keyword: str, has_tag_enabled=True):
        self.selector: Optional[Selector] = None
        self.function: Optional[Comparison] = None
        self.function_value = FunctionValue()

        self.keyword = keyword
        self.has_tag_enabled = has_tag_enabled

        self.not_true_reason = "A condition was not met."

    def load(self, id: str, data: dict):
        """ Parse the data into the conditional """
        self.selector = velvet_dawn.mechanics.selectors.get_selector(id, data[self.keyword])
        del data[self.keyword]

        if "notes" in data:
            del data["notes"]

        if "reason" in data:
            self.not_true_reason = data["reason"]
            del data['reason']

        for key in data:
            if key not in OPERATORS:
                raise errors.ValidationError(f"Invalid key in conditional: '{key}' in {id}.")

        if len(data) != 1:
            raise errors.ValidationError(f"Invalid operators in conditional: {data} in {id}. Conditionals must "
                                         f"contain only one operator.")

        # Test keys
        operator_key = list(data)[0]
        self.function = OPERATORS[operator_key]
        self.function_value.parse(id, data[operator_key])

        if self.function in NUMBERS_ONLY_OPERATORS:
            velvet_dawn.validations.is_int(self.function_value.raw, error_prefix=f"Conditional in {id} with operation '{operator_key}' comparison")

        if self.selector.attribute is not None and (self.function == Comparison.HAS_TAG or self.function == Comparison.NOT_HAS_TAG):
            raise errors.ValidationError(f"Cannot compare tags on this selector with an attribute on '{id}'.")
        if (self.function == Comparison.HAS_TAG or self.function == Comparison.NOT_HAS_TAG) and not self.has_tag_enabled:
            raise errors.ValidationError(f"Cannot compare tags using a '{self.keyword}' comparison method on '{id}'")

        return self

    def is_true(self, instance: Union[UnitInstance, TileInstance, WorldInstance]) -> bool:
        """ Call the conditional function """
        raise NotImplementedError()

File: src/jserver/src/velvetdawn/mechanics/conditionals/Conditional.java
package velvetdawn.mechanics.conditionals;

import velvetdawn.VelvetDawn;
import velvetdawn.mechanics.FunctionValue;
import velvetdawn.mechanics.selectors.Selector;
import velvetdawn.mechanics.selectors.Selectors;
import velvetdawn.models.anytype.Any;
import velvetdawn.models.anytype.AnyNull;
import velvetdawn.models.instances.Instance;
import velvetdawn.utils.Json;

import java.util.Map;
import java.util.Set;

public abstract class Conditional {

    // Comparison keys
    public static Map<String, Comparison> Operators = Map.of(
            "equals", Comparison.EQUALS,
            "eq", Comparison.EQUALS,
            "not-equals", Comparison.NOT_EQUALS,
            "ne", Comparison.NOT_EQUALS,
            "lt", Comparison.LESS_THAN,
            "lte", Comparison.LESS_THAN_EQUAL,
            "gt", Comparison.GREATER_THAN,
            "gte", Comparison.GREATER_THAN_EQUAL,
            "tagged", Comparison.HAS_TAG,
            "not-tagged", Comparison.NOT_HAS_TAG
    );

    // Numeric only operators
    public static Set<Comparison> NumberOnlyOperators = Set.of(
            Comparison.LESS_THAN,
            Comparison.LESS_THAN_EQUAL,
            Comparison.GREATER_THAN_EQUAL,
            Comparison.GREATER_THAN
    );

    protected Selector selector;
    public Comparison function;
    public FunctionValue functionValue;

    private final String keyword;
    private final boolean hasTagEnabled;

    public String notTrueReason = "A condition was not met.";

    /** Construct a conditional operator
     *
     * @param keyword The keyword to use for the conditional
     * @param hasTagEnabled Check if the comparison may compare tags.
     *                      This as not all conditional operations allow
     *                      you to compare against tags.
     */
    public Conditional(String keyword, boolean hasTagEnabled) {
        this.keyword = keyword;
        this.hasTagEnabled = hasTagEnabled;
    }

    public Conditional(String keyword) {
        this(keyword, true);
    }

    /** Parse the data into the conditional */
    public Conditional fromJson(VelvetDawn velvetDawn, String parentId, Json data) throws Exception {
        this.selector = Selectors.get(
                velvetDawn,
                parentId,
                data.get(this.keyword)
                                .validateInstanceIsString(String.format("Selectors must be strings (found in %s)", parentId))
                                .value);
        data.remove(this.keyword);

        data.remove("notes");

        this.notTrueReason = data.get("reason", Any.Null())
                .validateInstanceIsStringOrNull(String.format("Reason in conditional for %s must be a string", parentId))
                .toString();
        data.remove("reason");

        if (data.keys().size() != 1) {
            throw new Exception(String.format("Invalid operators in conditional: %s in %s. Conditionals must contain only one operator.", data.toString(), parentId));
        }

        for (String key: data.keys()) {
            if (!Operators.containsKey(key))
                throw new Exception(String.format("Invalid key in conditional: '%s' in %s.", key, parentId));
        }

        // Test keys
        String operatorKey = (String) data.keys().toArray()[0];
        this.function = Operators.get(operatorKey);
        this.functionValue = new FunctionValue(velvetDawn, parentId, data.get(operatorKey));

        if (NumberOnlyOperators.contains(this.function))
            this.functionValue.rawValue.validateInstanceIsFloat(String.format("Conditional in %s with operation '%s' comparison must be a float", parentId, operatorKey));

        // Can't compare tags if the condition explicitly looks to count number of items
        if (this.selector.attribute != null && (this.function == Comparison.HAS_TAG || this.function == Comparison.NOT_HAS_TAG))
            throw new Exception(String.format("Cannot compare tags on this selector with an attribute on '%s'.", parentId));
        if ((this.function == Comparison.HAS_TAG || this.function == Comparison.NOT_HAS_TAG) && !this.hasTagEnabled)
            throw new Exception(String.format("Cannot compare tags using a '{self.keyword}' comparison method on '%s'", parentId));

        return this;
    }

    /** Run the conditional function */
    public abstract boolean isTrue(Instance instance);
}
